core/src/main/scala/io/qbeast/spark/index/SparkPlanAnalyzer.scala

```diff
@@ -0,0 +1,109 @@
+package io.qbeast.spark.index
+
+import org.apache.spark.sql.catalyst.expressions.Alias
+import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.catalyst.plans.logical.Filter
+import org.apache.spark.sql.catalyst.plans.logical.LocalLimit
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.plans.logical.Project
+import org.apache.spark.sql.catalyst.plans.logical.Sample
+import org.apache.spark.sql.catalyst.plans.logical.Sort
+import org.apache.spark.sql.DataFrame
+
+import scala.collection.convert.ImplicitConversions.`collection asJava`
+
+object SparkPlanAnalyzer {
+
+  /**
+   * Builds a list of non-deterministic operations in the logical plan
+   *
+   * A list of Non-Deterministic operations:
+   *   - LocalLimit without Sort Operation
+   *   - Sample
+   *   - Filter with non-deterministic condition
+   *
+   * @param plan
+   *   the logical plan
+   * @return
+   */
+  private def isLogicalPlanDeterministic(plan: LogicalPlan): Boolean = {
+    // Recursively traverse the logical plan to find non-deterministic operations
+    val isCurrentOperationDeterministic = plan match {
+      case LocalLimit(_, _: Sort) => true // LocalLimit with Sort is deterministic
+      case LocalLimit(_, _) => false
+      case Sample(_, _, _, _, _) => false
```

Comment:
Doesn't this prevent a qbeast sample from being indexed with qbeast?
```
val qbeastDf = spark.read.format("qbeast").load(tablePath)
(qbeastDf
  .sample(0.1)
  .write
  .mode("append")
  .format("qbeast")
  .option("columnsToIndex", "col_1,col_2")
  .save(outputPath)
)
```

Reply:
You are right, although I had my concern at first..
 
Currently, the sample with Qbeast is deterministic, because we use the hash of the columns to reproduce the same weight. But is that premise going to be true in the future? Or should the Sample with Qbeast be deterministic by all means?

Reply:
Can you detect a qbeast sample?

Reply:
Yep, that is possible for sure